### ISSUE TITLE
Allow chaining conditionals

### ISSUE THREAD

**Issue by JohnstonJ** [casey/just#2411]
I wanted to write some code like this:

```
export VCPKG_DEFAULT_TRIPLET := if os_family() == "windows" && arch() == "x86_64" {
    "x64-windows-static"
} else if <snip> {
    <snip>
}
```

But apparently most common operators aren't implemented beyond what looks like string concatenation (judging from the error message).  For example, logical operators would be quite useful here.

```
error: Expected '{', '+', or '/', but found '&&'
 ——▶ justfile:1:61
  │
1 │ export VCPKG_DEFAULT_TRIPLET := if os_family() == "windows" && arch() == "x86_64" {
  │                                                             ^^
```

_This time_, I can work around this with a condition like `if os_family() + "-" + arch() == "windows-x86_64" {`, but I would imagine there are other scenarios out there where it doesn't work out as cleanly.

It seems that GNU make has similar limitations: https://stackoverflow.com/questions/6451477/makefile-ifeq-logical-and

But I don't know if those limitations from GNU make are worth keeping in just?  None of the proposed solutions in the above StackOverflow link look "better" to me than what the original poster was trying to do.

Some operators I could imagine being useful:
- Logical operators:
  - `&&`
  - `||`
  - `!` (unary NOT)
  - maybe an XOR operator
- Math operators: if the variable holds a number, then do some math on it. (`+`, `-`, `*`, `/`, etc).  But unless just gains a type system, IDK if that's something you'd want to do (else, `+` is ambiguous with strong concatenation)...

To me, the logical operators would be the most nice ones to have, but maybe keep options open for math operators in the future?

**Comment by casey** [comment]
Thanks for opening this issue!

Yah, conditionals are a massive hack. They exist as a special case in the grammar of conditionals, not as expressions. My motivation for implementing them in this terrible, awful, no good way, was because I've wanted to avoid committing to the representation of `true` and `false` in `just`. If logical operations are just allowed in conditionals, you can't get the result of evaluating one, so we don't need to decide what they evaluate too.

This is tricky, because I can imagine that one might want `&&` and `||` to behave in ways which aren't exactly the same as a logical operator that only takes `true` and `false`.

For example, you might want:

```
"a" || "" -> "a"
"" || "b" -> "b"
"a" || "b" -> "a"
"" || "" -> ""

"a" && "" -> ""
"" && "b" -> ""
"a" && "b" -> "b"
"" || "" -> ""
```

So `||` and `&&` behave as lazily-evaluated string-coalescing operators in expression contexts. I think this is actually the most natural, given that `just` is a string-oriented langauge, and this is what python `and` and `or` do.

For example, with the inbound PR #2440, you could do:

```
make := which("gmake") || which("make")
```

I.e., get make, preferring `gmake` if it exists, but otherwise falling back to `make.`

Whereas if `||` was a logical, this would evaluate to `true`/`false` or whatever the string representation of bools was.

**Comment by laniakea64** [comment]
> So `||` and `&&` behave as lazily-evaluated string-coalescing operators in expression contexts. I think this is actually the most natural,

Having the empty string be the "falsy" value wouldn't mix well with some of the existing syntaxes:
```
foo_path := "/something/that/evaluates/to/a/nonexistant/path"

# because path_exists() returns the string "false" for nonexistant paths,
# this would evaluate to foo_path
# but the most natural-seeming evaluation would be "/fallback"
foo := path_exists(foo_path) && foo_path || '/fallback'
```

Also, different justfiles may need different definitions of "falsy" (for example, the string `"0"` is truthy as an exit code stored in a backtick but falsy in a boolean `1`/`0` env var).

Maybe could let the user decide what strings are falsy on a per-justfile basis? -
```
set false-regex := '^\s*(?:[Ff]alse|0)\s*$'
```
To avoid surprises when using functions like `path_exists()`, this setting could default to `'^(false)$'` to match only empty strings and the string `"false"` by default.

**Comment by casey** [comment]
Yeah, I've ben thinking about our current functions that return `true` / `false`. They are `is_dependency`, `path_exists`, `semver_matches`.

My current thinking is that having them return `true` / `false` is actually just a mistake. Playing with `||` and `&&`, I think the empty string is just a much better false value.

I'm hesitant to introduce something like `false-regex`, because if users miss that setting, and mix `is_path` with `&&` the behavior they get is quite annoying and hard to figure out.

One option would be to deprecate those three functions, and introduce new versions which return `true` and the empty string, like, `is_recipe_dependency`, `does_path_exist` or (`exists`), and `is_semver_match`. These are not amazing names, but in this scenario, if you used the old versions, which didn't play nice with `&&` and `||`, you would get a warning, which seems less error prone.

I think this is something that's worth thinking hard about.

Just only has a single type, the string, and using strings to represent other types is not ideal, precisely because of issues like these. (An additional issue is the case where the empty string *is* a valid value, so doing something like `foo || fallback` can't be used in that case.)

I think using `"true"` / `"false"` is a mistake, just in the sense that the empty string is a better choice for false.

Another thought is that I've often thought that perhaps the base type should actually not be strings, but lists of strings. (See the [rc shell](https://www.scs.stanford.edu/nyu/04fa/sched/readings/rc.pdf), which does this.) In which case it wouldn't be the empty string which is the best representation of `false`, but the empty list.)

So, to stop rambling and sum up:

1. We could deprecate the three functions that return `true` and `false` and make new versions that return the empty string for `false.
2. We could add a setting like `false-regex` (control what `just` considers to be false) , or a setting like `empty-false` (make those three functions return the empty string for false)
3. We could write a static type system for `just`, so we can have more than one type.
4. I opened PR #2444 to add `&&` and `||`, and in lieu of figuring this out, I could land it as unstable, with a big warning in the readme documenting the unfortunate situation with the three functions which use `false` as the return value.

A static type system would be great, but it's a huge undertaking, and although I did take a compilers class in college where we learned about type systems and wrote a compiler for a statically-typed python-like language, that was a long time ago and I've basically forgotten everything 😅

Any other options I'm missing?

**Comment by laniakea64** [comment]
> 1. We could deprecate the three functions that return `true` and `false` and make new versions that return the empty string for `false.

That would be weird and seems unnecessary.  With such duplication, the use of the new functions represents an explicit opt-in to the new return value.  In which case there would be no need to have both the old and new functions simultaneously available for backwards compatibility, since justfiles using the new functions are updated anyway, and the functions would be identical in behavior all the way until the exact text of the true/false return values.

And as you pointed out, the new functions would need to have "not amazing" names, since the existing functions took the clear, concise names.  That the "better" functions have the worse names adds even more confusion.

> I could land it as unstable, with a big warning in the readme documenting the unfortunate situation with the three functions which use `false` as the return value.

Landing as unstable sounds like a good idea for at first :+1: regardless of the ultimate path forward.

> Any other options I'm missing?

5) Add a single boolean setting that:
 - is required to enable use of `&&` and `||` operators,
 - changes the "false" return values of the `just` functions to a falsy string,
 - allows conditionals without conditional operator, since strings can now be truthy/falsy:
```
foo := if path_exists("/some/path") { "It does exist" } else { "It doesn't exist" }
```

Not sure what would be a good name for this setting?

**Comment by casey** [comment]
> That would be weird and seems unnecessary. With such duplication, the use of the new functions represents an explicit opt-in to the new return value. In which case there would be no need to have both the old and new functions simultaneously available for backwards compatibility, since justfiles using the new functions are updated anyway, and the functions would be identical in behavior all the way until the exact text of the true/false return values.

Wouldn't you need both the new and the old functions? Maybe I'm misunderstanding, but removing the old functions would break old `justfile`s which use them.

> And as you pointed out, the new functions would need to have "not amazing" names, since the existing functions took the clear, concise names. That the "better" functions have the worse names adds even more confusion.

This is true, but a deprecation warning, along with removing the old functions from the docs would hopefully minimize that.

> Landing as unstable sounds like a good idea for at first 👍 regardless of the ultimate path forward.

I'll do that, just so I can get the PR in and then we can worry about the larger problems later.

> 5. Add a single boolean setting that:
> 
> * is required to enable use of `&&` and `||` operators,
> * changes the "false" return values of the `just` functions to a falsy string,
> * allows conditionals without conditional operator, since strings can now be truthy/falsy:
> 
> ```
> foo := if path_exists("/some/path") { "It does exist" } else { "It doesn't exist" }
> ```
> 
> Not sure what would be a good name for this setting?

I agree that the behavior of a setting like that is the end state that we want to get to, but I hesitate to add a setting like that because I think that you really just want it to be the default.

It would also mean that you would have to remember to add it to your `justfile` every time, and restricting the use of `&&` and `||`, changing the values of boolean functions, and allowing conditionals without an operator, would mean that in those places in the codebase, you would have to check what the setting was.

You could definitely do that, but it would be annoying. For example, the natural place to restrict `&&` and `||` is in the parser, but at that stage you don't know what the settings are, because you haven't parsed them yet, so you would have to defer checking the setting until the justfile was parsed.

Having worse names for three functions is lame, but seems less annoying and invasive to me.

**Comment by casey** [comment]
Landed `&&` and `||` as unstable in #2444.

**Comment by laniakea64** [comment]
> Wouldn't you need both the new and the old functions? Maybe I'm misunderstanding, but removing the old functions would break old `justfile`s which use them.

Settings are the usual answer for opting into backwards-incompatible behavior.  Having a setting to alter the functions' falsy return value would avoid duplication confusion and preserve the easy-to-understand existing function names.

With the only difference being in the string representation of one specific return value, it would seem the rationale that would favor deprecating instead of a setting would be if migrating to the new functions is onerous, i.e. justfiles may need the option of a transition period of gradually migrating the function calls instance-by-instance instead of doing it all at once.  But not seeing how that would be for only changing `"false"` to `""` :confused: 

> I hesitate to add a setting like that because I think that you really just want it to be the default.
> 
> It would also mean that you would have to remember to add it to your `justfile` every time, 

For what it's worth I had https://github.com/casey/just/issues/1201 in mind when thinking of that setting idea

> the natural place to restrict `&&` and `||` is in the parser, but at that stage you don't know what the settings are, because you haven't parsed them yet, so you would have to defer checking the setting until the justfile was parsed.
> 
> Having worse names for three functions is lame, but seems less annoying and invasive to me.

Oops, didn't realise the implementation impact :flushed: 

If the functions are going to be deprecated in favor of new ones, then instead of basically renaming, what about switching to non-boolean return values to prevent this problem recurring in face of a better (or different canonical) representation/idea of "truthy"?

This could reduce the negative impact of not being able to use the existing function names.

As examples of this idea:

- `path_exists(path)` -> `path_if_exists(path)`, returning the `path` argument unchanged if the path exists, and the empty string if the path does not exist.
- `is_dependency()` -> `depending_recipe()`, returning the name of the recipe that invoked this one as a dependency, or an empty string if the recipe is not a dependency.

For `semver_matches(version, req)` no idea how to adapt it along this line :fearful:

**Comment by casey** [comment]
I think I overstated the implementation impact. It's actually not so bad to note use of unstable features in the parser, and convert use of unstable features into errors in the analyzer later.

Let me rephrase I think the main reason I prefer deprecating the old functions. If the functions are deprecated and new versions with better behavior are introduced, the impact is:

1. Users who use the three old functions have to switch to the new functions.
2. The names of three functions will be slightly worse going forward.
3. Two version of the functions exist in the codebase.

1 is annoying, but it's a one-time cost, 2 is ongoing but I think pretty marginal, and 3 is ongoing but also pretty marginal.

Whereas with a setting:

1. If you want to use the new logical operators, you have to add a setting to your justfile.
2. The setting and different behaviors of the three functions need to be documented.

I think these are both slightly more annoying drawbacks than with deprecation, and they're ongoing, users will forever have to add the setting if they want to use the logical operators.

Not to overstate my case though! I think a setting isn't a terrible idea, and deprecation *is* annoying. You all of a sudden start getting warnings for things that used to work.

Rolling it into a hypothetical future edition does make it better, since multiple such changes can be bundled into a single setting, but it would still be better if we could avoid gating things behind editions, i.e., they're kind of a last-ditch solution for things that absolutely cannot be done in a backwards compatible way.

**Comment by casey** [comment]
It would be nice if this worked:

```just
just := `which just` && 'just' || just_executable()
```

The reason this doesn't work is because if `which just` fails, it halts execution instead of returning the empty string.

Maybe this isn't so bad, since you can do:

```just
just := `which just || true` && 'just' || just_executable()
```

**Comment by liquidaty** [comment]
*conditional* operators (e.g. `if a == 'b' *and* c == 'd' { ... } else { ... }`) do not need to be treated the same as *alternative* operators (e.g. `X := Y || Z`).

As a start, how about simply adding functions for each?

For conditionals: `and()` and `or()` (and optionally, `xor()` etc)? Note: this is a well-trodden path by e.g. Excel. This:
* eliminates parsing and compatibility complications
* is very easy to implement
* does not require any change to what "true" or "false" mean
  - assuming a bare conditional cannot be the right side of an assignment, which is no different from how conditionals currently work (where e.g. I cannot do `X := y == 'z'`)
* does not conflict with future implementations of other approaches discussed here

For alternatives: `coalesce(x,y,...)` for example.  I understand there has been [discussion of lazy evaluation](https://github.com/casey/just/issues/2379); however, it many (I might say, most) cases, lazy evaluation is neither needed nor desirable due to the potential confusion it causes. Furthermore, if lazy evaluation is desired, the scope of the topic is much broader than just coalesce-like features, and if it is to be addressed at all, would need some mechanism unrelated to this function.

I do think it's a good idea to make true/false values a 1st class expression (eg. that are assignable to vars), but the above would address pretty much all the use cases I can see in these discussions, without creating technical debt and while leaving all other options fully open

**Comment by casey** [comment]
What does `xor()` return when called with more than two arguments?

Adding `and()` and `or()` isn't totally unreasonable, although I think I'd be unfortunate , given that they duplicate `||` and `&&`. It would be confusing to have two ways of doing the same thing. I'd really like to resolve the true false thing, and just make boolean operations first class.

**Comment by liquidaty** [comment]
> What does `xor()` return when called with more than two arguments?

Could be the same behavior as: https://support.microsoft.com/en-us/office/xor-function-1548d4c2-5e47-4f77-9a92-0533bba14f37

> Adding `and()` and `or()` isn't totally unreasonable, although I think I'd be unfortunate , given that they duplicate `||` and `&&`. It would be confusing to have two ways of doing the same thing. I'd really like to resolve the true false thing, and just make boolean operations first class.

Personally, I agree. Then again, using functions is familiar to many and easy to learn, and easy to implement and maintain, so I don't see much downside and some users may actually prefer, at least it seems that way given that all spreadsheets support it-- and Excel has an estimated 500mm- 1Bn users.-- as does e.g. `numpy` (`logical_and` ...), Python's `operator` module etc

**Comment by casey** [comment]
Implemented in #3387. Requires the unstable `set lists` setting, because we use the empty list, `[]`, to represent false.
